tests/tensor/test_nlinalg.py

```diff
@@ -120,6 +123,117 @@ def test_matrix_dot():
     assert _allclose(numpy_sol, aesara_sol)
 
 
+class TestMatMul(utt.InferShapeTester):
+    def setup_method(self):
+        super().setup_method()
+        self.rng = np.random.default_rng(utt.fetch_seed())
+        self.op = matmul
+        self.op_class = MatMul
+
+    def _validate_output(self, a, b):
+        aesara_sol = self.op(a, b).eval()
+        numpy_sol = np.matmul(a, b)
+        assert _allclose(numpy_sol, aesara_sol)
+
+    def test_op(self):
+        # test output when both inputs are vectors
+        self._validate_output(
+            self.rng.random(3).astype(config.floatX),
+            self.rng.random(3).astype(config.floatX),
+        )
+
+        # test output when both inputs are matrices
+        x1 = self.rng.random((5, 3)).astype(config.floatX)
+        x2 = self.rng.random((3, 2)).astype(config.floatX)
+        self._validate_output(x1, x2)
+
+        # test behaviour when one of the inputs is has dimension > 2
+        x2 = self.rng.random((5, 10, 1, 3, 5)).astype(config.floatX)
+        self._validate_output(x1, x2)
+        self._validate_output(x2, x1)
+
+        # test behaviour when one of the inputs is a vector
+        x2 = self.rng.random(3).astype(config.floatX)
+        self._validate_output(x1, x2)
+        x2 = self.rng.random(5).astype(config.floatX)
+        self._validate_output(x2, x1)
+
+        # check if behaviour is correct N-D arrays where N > 2.
+        x1 = np.arange(2 * 2 * 4).reshape((2, 2, 4)).astype(config.floatX)
+        x2 = np.arange(2 * 2 * 4).reshape((2, 4, 2)).astype(config.floatX)
+        self._validate_output(x1, x2)
+
+        with pytest.raises(ValueError, match="cannot be scalar"):
+            self.op(4, [4, 1])
+
+    def test_dtype_param(self):
+        for dtype in (np.float16, np.float32, np.float64):
+            sol = self.op([1, 2, 3], [3, 2, 1], dtype=dtype)
+            assert sol.eval().dtype == dtype
+
+    @pytest.mark.parametrize(
+        "x1_shape,x2_shape,error_regex",
+        [
+            ((1,), (3,), "inputs must have the same length"),
+            ((2,), (3, 1), "length of input 1.*2nd-last dimension of input 2"),
+            ((2, 5), (3,), "length of input 2.*of the last dimension of input 1"),
+            (
+                (2, 5),
+                (3, 4),
+                "number of columns of input 1 .* number of rows of input 2",
+            ),
+            (
+                (2, 1, 3),
+                (5, 4),
+                "number of rows of input 2 .* last dimension of input 1",
+            ),
+            (
+                (2, 5),
+                (2, 4, 3),
+                "number of columns of input 1 .* 2nd-last dimension of input 2",
+            ),
+            (
+                (3, 2, 4, 5),
+                (1, 6, 7),
+                "length of the last dimension of input 1 .* 2nd-last dimension of input 2",
+            ),
+            (
+                (4, 5, 8, 2, 3),
+                (3, 2, 3, 2, 5, 8, 3, 3),
```

Comment:
Do these test arrays need to be this large in order to reach the relevant code path?